// @storybook/ui/dist/libs/withLifecycleDecorator.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

// A small utility to add before/afterEach to stories.
var WithLifecyle =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(WithLifecyle, _Component);

  function WithLifecyle(props) {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, WithLifecyle);

    for (var _len = arguments.length, rest = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
      rest[_key - 1] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(WithLifecyle)).call.apply(_getPrototypeOf2, [this, props].concat(rest)));
    props.beforeEach();
    return _this;
  }

  (0, _createClass2.default)(WithLifecyle, [{
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      var afterEach = this.props.afterEach;
      afterEach();
    }
  }, {
    key: "render",
    value: function render() {
      var storyFn = this.props.storyFn;
      return storyFn();
    }
  }]);
  return WithLifecyle;
}(_react.Component);

WithLifecyle.propTypes = {
  storyFn: _propTypes.default.func.isRequired,
  beforeEach: _propTypes.default.func,
  afterEach: _propTypes.default.func
};
WithLifecyle.defaultProps = {
  beforeEach: function beforeEach() {},
  afterEach: function afterEach() {}
};

var _default = function _default(_ref) {
  var beforeEach = _ref.beforeEach,
      afterEach = _ref.afterEach;
  return function (storyFn) {
    return _react.default.createElement(WithLifecyle, {
      beforeEach: beforeEach,
      afterEach: afterEach,
      storyFn: storyFn
    });
  };
};

exports.default = _default;
