// @storybook/ui/dist/modules/shortcuts/actions/shortcuts.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.keyEventToOptions = keyEventToOptions;
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _objectSpread3 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _pick = _interopRequireDefault(require("lodash/pick"));

var _key_events = require("../../../libs/key_events");

var _actions = _interopRequireDefault(require("../../api/actions"));

var _defaultState = _interopRequireDefault(require("../defaultState"));

var deprecationMessage = function deprecationMessage(oldName, newName) {
  return "The ".concat(oldName, " option has been renamed to ").concat(newName, " and will not be available in the next major Storybook release. Please update your config.");
};

function keyEventToOptions(currentOptions, event) {
  if (currentOptions.enableShortcuts) {
    switch (event) {
      case _key_events.features.FULLSCREEN:
        return {
          goFullScreen: !currentOptions.goFullScreen
        };

      case _key_events.features.ADDON_PANEL:
        return {
          showAddonPanel: !currentOptions.showAddonPanel
        };

      case _key_events.features.STORIES_PANEL:
        return {
          showStoriesPanel: !currentOptions.showStoriesPanel
        };

      case _key_events.features.SHOW_SEARCH:
        return {
          showSearchBox: true
        };

      case _key_events.features.ADDON_PANEL_IN_RIGHT:
        return {
          addonPanelInRight: !currentOptions.addonPanelInRight
        };

      default:
        return {};
    }
  }

  return {};
}

var renamedOptions = {
  showLeftPanel: 'showStoriesPanel',
  showDownPanel: 'showAddonPanel',
  downPanelInRight: 'addonPanelInRight'
};
var _default = {
  handleEvent: function handleEvent(context, event) {
    var clientStore = context.clientStore;

    switch (event) {
      case _key_events.features.NEXT_STORY:
        _actions.default.api.jumpToStory(context, 1);

        break;

      case _key_events.features.PREV_STORY:
        _actions.default.api.jumpToStory(context, -1);

        break;

      default:
        clientStore.update(function (state) {
          var newOptions = keyEventToOptions(state.shortcutOptions, event);
          var updatedOptions = (0, _objectSpread3.default)({}, state.shortcutOptions, newOptions);
          return {
            shortcutOptions: updatedOptions
          };
        });
    }
  },
  setOptions: function setOptions(_ref, options) {
    var clientStore = _ref.clientStore;
    clientStore.update(function () {
      var updatedOptions = (0, _objectSpread3.default)({}, _defaultState.default.shortcutOptions, (0, _pick.default)(options, Object.keys(_defaultState.default.shortcutOptions)));
      var withNewNames = Object.keys(renamedOptions).reduce(function (acc, oldName) {
        var newName = renamedOptions[oldName];

        if (oldName in options && !(newName in options)) {
          if (process.env.NODE_ENV !== 'production') {
            // eslint-disable-next-line no-console
            console.warn(deprecationMessage(oldName, newName));
          }

          return (0, _objectSpread3.default)({}, acc, (0, _defineProperty2.default)({}, newName, options[oldName]));
        }

        return acc;
      }, updatedOptions);
      return {
        shortcutOptions: withNewNames
      };
    });
  }
};
exports.default = _default;
