// @storybook/ui/dist/modules/ui/configs/handle_routing.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getUrlState = getUrlState;
exports.changeUrl = changeUrl;
exports.updateStore = updateStore;
exports.handleInitialUrl = handleInitialUrl;
exports.default = _default;
exports.config = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _global = require("global");

var _qs = _interopRequireDefault(require("qs"));

var config = {
  insidePopState: false
};
exports.config = config;

function getUrlState(data) {
  var selectedKind = data.selectedKind,
      selectedStory = data.selectedStory,
      customQueryParams = data.customQueryParams;
  var _data$shortcutOptions = data.shortcutOptions,
      full = _data$shortcutOptions.goFullScreen,
      addons = _data$shortcutOptions.showAddonPanel,
      stories = _data$shortcutOptions.showStoriesPanel,
      panelRight = _data$shortcutOptions.addonPanelInRight;
  var addonPanel = data.selectedAddonPanel;
  var urlObj = (0, _objectSpread2.default)({
    selectedKind: selectedKind,
    selectedStory: selectedStory,
    full: Number(full),
    addons: Number(addons),
    stories: Number(stories),
    panelRight: Number(panelRight),
    addonPanel: addonPanel
  }, customQueryParams);
  var url = "?".concat(_qs.default.stringify(urlObj));
  return (0, _objectSpread2.default)({}, urlObj, {
    full: full,
    addons: addons,
    stories: stories,
    panelRight: panelRight,
    url: url
  });
}

function changeUrl(clientStore, usePush) {
  // Do not change the URL if we are inside a popState event.
  if (config.insidePopState) return;
  var data = clientStore.getAll();
  if (!data.selectedKind) return;
  var state = getUrlState(data);

  try {
    _global.history[usePush ? 'pushState' : 'replaceState'](state, '', state.url);
  } catch (e) {// do nothing
  }
}

function updateStore(queryParams, actions) {
  var selectedKind = queryParams.selectedKind,
      selectedStory = queryParams.selectedStory,
      _queryParams$full = queryParams.full,
      full = _queryParams$full === void 0 ? 0 : _queryParams$full,
      _queryParams$down = queryParams.down,
      down = _queryParams$down === void 0 ? 1 : _queryParams$down,
      _queryParams$addons = queryParams.addons,
      addons = _queryParams$addons === void 0 ? down : _queryParams$addons,
      _queryParams$left = queryParams.left,
      left = _queryParams$left === void 0 ? 1 : _queryParams$left,
      _queryParams$stories = queryParams.stories,
      stories = _queryParams$stories === void 0 ? left : _queryParams$stories,
      _queryParams$panelRig = queryParams.panelRight,
      panelRight = _queryParams$panelRig === void 0 ? 0 : _queryParams$panelRig,
      downPanel = queryParams.downPanel,
      _queryParams$addonPan = queryParams.addonPanel,
      addonPanel = _queryParams$addonPan === void 0 ? downPanel : _queryParams$addonPan,
      customQueryParams = (0, _objectWithoutProperties2.default)(queryParams, ["selectedKind", "selectedStory", "full", "down", "addons", "left", "stories", "panelRight", "downPanel", "addonPanel"]);

  if (selectedKind) {
    actions.api.selectStory(selectedKind, selectedStory);
  }

  actions.shortcuts.setOptions({
    goFullScreen: Boolean(Number(full)),
    showAddonPanel: Boolean(Number(addons)),
    showStoriesPanel: Boolean(Number(stories)),
    addonPanelInRight: Boolean(Number(panelRight))
  });

  if (addonPanel) {
    actions.ui.selectAddonPanel(addonPanel);
  }

  actions.api.setQueryParams(customQueryParams);
}

function handleInitialUrl(actions, l) {
  var queryString = l.search.substring(1);
  if (!queryString || queryString === '') return;

  var parsedQs = _qs.default.parse(queryString);

  updateStore(parsedQs, actions);
}

function _default(_ref, actions) {
  var clientStore = _ref.clientStore;
  // handle initial URL
  handleInitialUrl(actions, _global.location);
  var data = clientStore.getAll();
  var prevKind = data.selectedKind;
  var prevStory = data.selectedStory; // subscribe to clientStore and change the URL

  clientStore.subscribe(function () {
    var _clientStore$getAll = clientStore.getAll(),
        selectedKind = _clientStore$getAll.selectedKind,
        selectedStory = _clientStore$getAll.selectedStory; // use pushState only when a new story is selected


    var usePush = prevKind != null && prevStory != null && (selectedKind !== prevKind || selectedStory !== prevStory);
    changeUrl(clientStore, usePush);
    prevKind = selectedKind;
    prevStory = selectedStory;
  });
  changeUrl(clientStore); // handle back button

  _global.window.onpopstate = function () {
    config.insidePopState = true;
    handleInitialUrl(actions, _global.location);
    config.insidePopState = false;
  };
}
