// @storybook/addon-links/dist/react/components/link.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _components = require("@storybook/components");

var _preview = require("../../preview");

var LinkTo =
/*#__PURE__*/
function (_PureComponent) {
  (0, _inherits2.default)(LinkTo, _PureComponent);

  function LinkTo() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, LinkTo);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(LinkTo)).call.apply(_getPrototypeOf2, [this].concat(args)));
    _this.state = {
      href: '/'
    };
    _this.handleClick = _this.handleClick.bind((0, _assertThisInitialized2.default)((0, _assertThisInitialized2.default)(_this)));
    return _this;
  }

  (0, _createClass2.default)(LinkTo, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.updateHref();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      var _this$props = this.props,
          kind = _this$props.kind,
          story = _this$props.story;

      if (prevProps.kind !== kind || prevProps.story !== story) {
        this.updateHref();
      }
    }
  }, {
    key: "updateHref",
    value: function () {
      var _updateHref = (0, _asyncToGenerator2.default)(
      /*#__PURE__*/
      _regenerator.default.mark(function _callee() {
        var _this$props2, kind, story, href;

        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _this$props2 = this.props, kind = _this$props2.kind, story = _this$props2.story;
                _context.next = 3;
                return (0, _preview.hrefTo)(kind, story);

              case 3:
                href = _context.sent;
                this.setState({
                  href: href
                });

              case 5:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      return function updateHref() {
        return _updateHref.apply(this, arguments);
      };
    }()
  }, {
    key: "handleClick",
    value: function handleClick() {
      (0, _preview.openLink)(this.props);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props3 = this.props,
          kind = _this$props3.kind,
          story = _this$props3.story,
          rest = (0, _objectWithoutProperties2.default)(_this$props3, ["kind", "story"]);
      var href = this.state.href;
      return _react.default.createElement(_components.RoutedLink, (0, _extends2.default)({
        href: href,
        onClick: this.handleClick
      }, rest));
    }
  }]);
  return LinkTo;
}(_react.PureComponent);

exports.default = LinkTo;
LinkTo.defaultProps = {
  kind: null,
  story: null
};
LinkTo.propTypes = {
  kind: _propTypes.default.string,
  story: _propTypes.default.string
};
