// @storybook/core/dist/server/config/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loadEnv = loadEnv;
exports.getBabelRuntimePath = exports.nodeModulesPaths = exports.excludePaths = exports.includePaths = void 0;

var _path = _interopRequireDefault(require("path"));

var _lazyUniversalDotenv = require("lazy-universal-dotenv");

const includePaths = [_path.default.resolve('./')];
exports.includePaths = includePaths;
const excludePaths = [_path.default.resolve('node_modules')];
exports.excludePaths = excludePaths;

const nodeModulesPaths = _path.default.resolve('./node_modules');

exports.nodeModulesPaths = nodeModulesPaths;

const nodePathsToArray = nodePath => nodePath.split(process.platform === 'win32' ? ';' : ':').filter(Boolean).map(p => _path.default.resolve('./', p)); // Load environment variables starts with STORYBOOK_ to the client side.


function loadEnv(options = {}) {
  const defaultNodeEnv = options.production ? 'production' : 'development';
  const env = {
    NODE_ENV: process.env.NODE_ENV || defaultNodeEnv,
    NODE_PATH: process.env.NODE_PATH || '',
    // This is to support CRA's public folder feature.
    // In production we set this to dot(.) to allow the browser to access these assests
    // even when deployed inside a subpath. (like in GitHub pages)
    // In development this is just empty as we always serves from the root.
    PUBLIC_URL: options.production ? '.' : ''
  };
  Object.keys(process.env).filter(name => /^STORYBOOK_/.test(name)).forEach(name => {
    env[name] = process.env[name];
  });
  const base = Object.entries(env).reduce((acc, [k, v]) => Object.assign(acc, {
    [k]: JSON.stringify(v)
  }), {});
  const {
    stringified,
    raw
  } = (0, _lazyUniversalDotenv.getEnvironment)({
    nodeEnv: env.NODE_ENV
  });
  const fullRaw = Object.assign({}, env, raw);
  fullRaw.NODE_PATH = nodePathsToArray(fullRaw.NODE_PATH || '');
  return {
    stringified: Object.assign({}, base, stringified),
    raw: fullRaw
  };
}

const getBabelRuntimePath = () => {
  const pkgJsonPath = require.resolve('@babel/runtime/package.json');

  return _path.default.dirname(pkgJsonPath);
};

exports.getBabelRuntimePath = getBabelRuntimePath;
