// @storybook/addon-actions/dist/preview/action.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = action;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _v = _interopRequireDefault(require("uuid/v1"));

var _addons = _interopRequireDefault(require("@storybook/addons"));

var _constants = require("../constants");

var _util = require("../lib/util");

var _configureActions = require("./configureActions");

function action(name) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var actionOptions = (0, _objectSpread2.default)({}, _configureActions.config, options); // eslint-disable-next-line no-shadow

  var handler = function action() {
    for (var _len = arguments.length, _args = new Array(_len), _key = 0; _key < _len; _key++) {
      _args[_key] = arguments[_key];
    }

    var args = _args.map(function (arg) {
      return (0, _util.prepareArguments)(arg, actionOptions.depth);
    });

    var channel = _addons.default.getChannel();

    var id = (0, _v.default)();
    channel.emit(_constants.EVENT_ID, {
      id: id,
      data: {
        name: name,
        args: args
      },
      options: actionOptions
    });
  };

  if (_util.canConfigureName && name && typeof name === 'string') {
    Object.defineProperty(handler, 'name', {
      value: name
    });
  }

  return handler;
}
